Problem set 2: Aggregation, reshaping, and validation

Due: 11:59pm on Friday, September 13 by uploading to Brightspace

Your name:

1. For this problem set, use any of the following datasets:

* [NOAA](https://www.ncei.noaa.gov/products) weather and climate data
* [World Happiness Report](https://worldhappiness.report/ed/2021/#appendices-and-data)
* [NASA data](https://data.nasa.gov/)
* [US Census data](https://data.census.gov/)
* [WHO COVID data](https://covid19.who.int/)
* [Netflix data](https://www.kaggle.com/datasets/ashishgup/netflix-rotten-tomatoes-metacritic-imdb)
* [Nashville open data](https://data.nashville.gov/)
* [Yelp reviews data](https://www.yelp.com/dataset)

In [ ]:
# Read in the data
import pandas as pd
df = pd.read_excel('./WHR21_Data_Mortality.xlsx')

# Do some investigative EDA to identify the structure of the data and its content. Then write a brief summary of the dataset in your own words.
print(df.info())

# Rename columns for easier access
df = df.rename(columns={
    "Country name": "country_name",
    "Population 2020": "population_2020",
    "Population 2019": "population_2019",
    "COVID-19 deaths per 100,000 population in 2020": "covid19_deaths_per_100k_2020",
    "Median age": "median_age",
    "Island": "island",
    "Index of exposure to COVID-19  infections in other countries as of March 31": "exposure_index_march31",
    "Log of average distance to SARS countries": "log_avg_distance_sars_countries",
    "WHO Western Pacific Region": "who_western_pacific_region",
    "Female head of government": "female_head_of_government",
    "Index of institutional trust": "institutional_trust_index",
    "Gini coefficient of income": "gini_coefficient",
    "All-cause death count, 2017": "all_cause_deaths_2017",
    "All-cause death count, 2018": "all_cause_deaths_2018",
    "All-cause death count, 2019": "all_cause_deaths_2019",
    "All-cause death count, 2020": "all_cause_deaths_2020",
    "Excess deaths in 2020 per 100,000 population, relative to 2017-2019 average": "excess_deaths_per_100k_2020"
})

print(df.head())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 166 entries, 0 to 165
Data columns (total 17 columns):
 #   Column                                                                       Non-Null Count  Dtype  
---  ------                                                                       --------------  -----  
 0   Country name                                                                 166 non-null    object 
 1   Population 2020                                                              164 non-null    float64
 2   Population 2019                                                              164 non-null    float64
 3   COVID-19 deaths per 100,000 population in 2020                               163 non-null    float64
 4   Median age                                                                   163 non-null    float64
 5   Island                                                                       166 non-null    int64  
 6   Index of exposure to COVID-19  infections 

In [2]:
display(df.sample(10))

country_name  population_2020  population_2019  \
119                    Jamaica     2.961161e+06     2.948279e+06   
100                    Croatia     4.105268e+06     4.067500e+06   
26   Hong Kong S.A.R. of China     7.496988e+06     7.451000e+06   
160                 Uzbekistan     3.346920e+07     3.358065e+07   
30                       India     1.380004e+09     1.366418e+09   
95                       Chile     1.911621e+07     1.895204e+07   
59                  Mauritania     4.649660e+06     4.525696e+06   
134                    Namibia     2.540916e+06     2.494530e+06   
10                  Bangladesh     1.646894e+08     1.630462e+08   
52                        Laos     7.275556e+06     7.169455e+06   

     covid19_deaths_per_100k_2020  median_age  island  exposure_index_march31  \
119                     10.198700   31.400000       1                1.044930   
100                     95.487099   44.000000       0                4.057454   
26                       1.974126   44.799999       0                1.178745   
160                      1.834500   28.200001       0                1.140722   
30                      10.778100   28.200001       0                0.936681   
95                      86.879196   35.400002       0                0.562487   
59                       7.462900   20.299999       0                1.149162   
134                      8.068000   22.000000       0                0.625211   
10                       4.589900   27.500000       0                0.823931   
52                       0.000000   24.400000       0                0.772761   

     log_avg_distance_sars_countries  who_western_pacific_region  \
119                         9.502089                           0   
100                         9.037495                           0   
26                          8.044208                           1   
160                         8.620495                           0   
30                          8.532171                           0   
95                          9.722240                           0   
59                          9.363724                           0   
134                         9.349452                           0   
10                          8.348938                           0   
52                          8.183916                           1   

     female_head_of_government  institutional_trust_index  gini_coefficient  \
119                          0                   0.133247         37.480000   
100                          0                   0.131280         29.700000   
26                           0                   0.317925         48.275000   
160                          0                   0.434861         39.700000   
30                           0                   0.526049         39.334375   
95                           0                   0.096694         45.920000   
59                           0                   0.122432         32.620000   
134                          1                   0.310684         57.535000   
10                           1                   0.576998         43.972500   
52                           0                   0.707731         36.390000   

     all_cause_deaths_2017  all_cause_deaths_2018  all_cause_deaths_2019  \
119                    NaN                    NaN                    NaN   
100                53269.0                52562.0                51611.0   
26                 45883.0                47478.0                48706.0   
160               160723.0               154913.0               154570.0   
30                     NaN                    NaN                    NaN   
95                106045.0               106458.0               109079.0   
59                     NaN                    NaN                    NaN   
134                    NaN                    NaN                    NaN   
10                     NaN                    NaN                    NaN  

### Summary
*This dataset, from World Hapiness Report, contains information on 166 countries, including their demographics (such as population size and age distribution), pandemic data (COVID-19 deaths and excess mortality), structural features (island status, exposure to international spread), and social-economic factors (institutional trust, income inequality, female leadership) in 2020. Most data are well documented, while records of all-cause mortality from previous years are less complete.*


2. Aggregate and reshape

In [3]:
# Reshape the data in some way that changes the unit of analysis. Convert it into an aggregated summary table that provides new information from that table
group_by_island = df.groupby("island").agg({
    "population_2020": "mean",
    "median_age": "mean",
    "covid19_deaths_per_100k_2020": "mean",
    "excess_deaths_per_100k_2020": "mean",
    "institutional_trust_index": "mean"
}).reset_index()

group_by_island
# Write 3-4 sentences summarizing what new insight we get from the aggregated table


island  population_2020  median_age  covid19_deaths_per_100k_2020  \
0       0     4.947931e+07   30.101418                     33.172511   
1       1     3.234825e+07   34.295455                     14.883223   

   excess_deaths_per_100k_2020  institutional_trust_index  
0                   119.959358                   0.308743  
1                    43.188946                   0.369632

### Summary

*From the aggregated data, we see island nations tend to have smaller populations and  older median ages compared to non-island countries. They had lower COVID-19 deaths and excess mortality rates, potentially due to their geographic isolation. Island countries also show somewhat higher institutional trust.*


3. Validation and missingness


In [4]:
# Choose one numeric column from either one of the original tables or your newly merged table. Convert a random 5% sample of values to NaN. Make sure you set a seed first so the dropped values are always the same
import numpy as np

np.random.seed(42)

df_num = df.copy()

random_indices = np.random.choice(df_num.index, size=int(0.05 * len(df_num)), replace=False)
df_num.loc[random_indices, "COVID-19 deaths per 100,000 population in 2020"] = np.nan

# Calculate the percent of missing values in each column

missing_percent = df_num.isnull().mean() * 100
print("Percent of missing values in each column:")
print(missing_percent)

# Find the mean of the numeric column

mean_value = df_num["COVID-19 deaths per 100,000 population in 2020"].mean()
print(f"Mean of 'COVID-19 deaths per 100,000 population in 2020': {mean_value}")

# Use listwise deletion and find the mean again

mean_value_listwise = df_num["COVID-19 deaths per 100,000 population in 2020"].dropna().mean()
print(f"Mean of 'COVID-19 deaths per 100,000 population in 2020' (listwise deletion): {mean_value_listwise}")

# Use mean imputation and find the mean again

df_num["COVID-19 deaths per 100,000 population in 2020"].fillna(mean_value, inplace=True)
mean_value_imputed = df_num["COVID-19 deaths per 100,000 population in 2020"].mean()
print(f"Mean of 'COVID-19 deaths per 100,000 population in 2020' (mean imputation): {mean_value_imputed}")

Percent of missing values in each column:
country_name                                        0.000000
population_2020                                     1.204819
population_2019                                     1.204819
covid19_deaths_per_100k_2020                        1.807229
median_age                                          1.807229
island                                              0.000000
exposure_index_march31                              1.204819
log_avg_distance_sars_countries                     1.204819
who_western_pacific_region                          0.000000
female_head_of_government                           0.000000
institutional_trust_index                           0.602410
gini_coefficient                                    1.204819
all_cause_deaths_2017                              62.048193
all_cause_deaths_2018                              61.445783
all_cause_deaths_2019                              61.445783
all_cause_deaths_2020                      

C:\temp\ipykernel_8240\2331257004.py:29: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_num["COVID-19 deaths per 100,000 population in 2020"].fillna(mean_value, inplace=True)


In [5]:
# Use a more advanced form of imputation and find the mean again
df_num = df.copy()
random_indices = np.random.choice(df_num.index, size=int(0.05 * len(df_num)), replace=False)
df_num.loc[random_indices, "COVID-19 deaths per 100,000 population in 2020"] = np.nan

# Median imputation
median_value = df_num["COVID-19 deaths per 100,000 population in 2020"].median()
df_num_median = df_num.copy()
df_num_median["COVID-19 deaths per 100,000 population in 2020"].fillna(median_value, inplace=True)

mean_value_median = df_num_median["COVID-19 deaths per 100,000 population in 2020"].mean()
print(f"Mean of 'COVID-19 deaths per 100,000 population in 2020' (median imputation): {mean_value_median}")


Mean of 'COVID-19 deaths per 100,000 population in 2020' (median imputation): nan


C:\temp\ipykernel_8240\4147809403.py:9: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_num_median["COVID-19 deaths per 100,000 population in 2020"].fillna(median_value, inplace=True)


### Compare these 4 values and in 3-4 sentences decribe that it tells us about imputation and missing data

*When comparing the four approaches, the values are close but not equal. Listwise deletion reduces the sample size and can slightly bias the result if the missing values are not random. Mean imputation keeps the mean stable but reduces the variance. Median imputation is less sensitive to outliers.*

In [6]:
# Repeat the above steps with a factor/string column, this time doing mode imputation instead of mean
df_cat = df.copy()
np.random.seed(42)

random_indices = np.random.choice(df_cat.index, size=int(0.05 * len(df_cat)), replace=False)
df_cat.loc[random_indices, "female_head_of_government"] = np.nan

missing_percent_cat = df_cat.isnull().mean() * 100
print("Percent of missing values in each column:")
print(missing_percent_cat)

# Original mode
true_mode = df["female_head_of_government"].mode()[0]

# Listwise deletion
listwise_mode = df_cat["female_head_of_government"].dropna().mode()[0]

# Mode imputation
mode_value = df_cat["female_head_of_government"].mode()[0]
df_cat_mode = df_cat.copy()
df_cat_mode["female_head_of_government"].fillna(mode_value, inplace=True)
mode_imputed = df_cat_mode["female_head_of_government"].mode()[0]

print(f"Original mode: {true_mode}")
print(f"Mode after listwise deletion: {listwise_mode}")
print(f"Mode after mode imputation: {mode_imputed}")

Percent of missing values in each column:
country_name                        0.000000
population_2020                     1.204819
population_2019                     1.204819
covid19_deaths_per_100k_2020        1.807229
median_age                          1.807229
island                              0.000000
exposure_index_march31              1.204819
log_avg_distance_sars_countries     1.204819
who_western_pacific_region          0.000000
female_head_of_government           4.819277
institutional_trust_index           0.602410
gini_coefficient                    1.204819
all_cause_deaths_2017              62.048193
all_cause_deaths_2018              61.445783
all_cause_deaths_2019              61.445783
all_cause_deaths_2020              60.843373
excess_deaths_per_100k_2020        61.445783
dtype: float64
Original mode: 0
Mode after listwise deletion: 0.0
Mode after mode imputation: 0.0


C:\temp\ipykernel_8240\1103400183.py:21: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_cat_mode["female_head_of_government"].fillna(mode_value, inplace=True)
